modules/store/src/provide_store.ts

```diff
@@ -0,0 +1,343 @@
+import {
+  Inject,
+  InjectionToken,
+  Injector,
+  Optional,
+  SkipSelf,
+  ImportedNgModuleProviders,
+  ENVIRONMENT_INITIALIZER,
+  inject,
+  InjectFlags,
+  Provider,
+} from '@angular/core';
+import {
+  Action,
+  ActionReducer,
+  ActionReducerMap,
+  StoreFeature,
+} from './models';
+import { combineReducers, createReducerFactory } from './utils';
+import {
+  INITIAL_STATE,
+  INITIAL_REDUCERS,
+  _INITIAL_REDUCERS,
+  REDUCER_FACTORY,
+  _REDUCER_FACTORY,
+  STORE_FEATURES,
+  _INITIAL_STATE,
+  META_REDUCERS,
+  _STORE_REDUCERS,
+  FEATURE_REDUCERS,
+  _FEATURE_REDUCERS,
+  _FEATURE_REDUCERS_TOKEN,
+  _STORE_FEATURES,
+  _FEATURE_CONFIGS,
+  USER_PROVIDED_META_REDUCERS,
+  _RESOLVED_META_REDUCERS,
+  _ROOT_STORE_GUARD,
+  _ACTION_TYPE_UNIQUENESS_CHECK,
+} from './tokens';
+import { ACTIONS_SUBJECT_PROVIDERS, ActionsSubject } from './actions_subject';
+import {
+  REDUCER_MANAGER_PROVIDERS,
+  ReducerManager,
+  ReducerObservable,
+} from './reducer_manager';
+import {
+  SCANNED_ACTIONS_SUBJECT_PROVIDERS,
+  ScannedActionsSubject,
+} from './scanned_actions_subject';
+import { STATE_PROVIDERS } from './state';
+import { STORE_PROVIDERS, Store } from './store';
+import {
+  provideRuntimeChecks,
+  checkForActionTypeUniqueness,
+} from './runtime_checks';
+import {
+  FeatureSlice,
+  RootStoreConfig,
+  StoreConfig,
+  _concatMetaReducers,
+  _createFeatureReducers,
+  _createFeatureStore,
+  _createStoreReducers,
+  _initialStateFactory,
+  _provideForRootGuard,
+} from './store_config';
+
+/**
+ * InjectionToken that registers feature states.
+ * Mainly used to provide a hook that can be injected
+ * to ensure feature state is loaded before something
+ * that depends on it.
+ */
+export const FEATURE_STATE_PROVIDER = new InjectionToken('NgRx Feature State', {
+  factory() {
+    const features = inject<StoreFeature<any, any>[]>(_STORE_FEATURES);
+    const featureReducers = inject<ActionReducerMap<any>[]>(FEATURE_REDUCERS);
+    const reducerManager = inject(ReducerManager);
+    inject(_ACTION_TYPE_UNIQUENESS_CHECK, InjectFlags.Optional);
+
+    const feats = features.map((feature, index) => {
+      const featureReducerCollection = featureReducers.shift();
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const reducers = featureReducerCollection! /*TODO(#823)*/[index];
+
+      return {
+        ...feature,
+        reducers,
+        initialState: _initialStateFactory(feature.initialState),
+      };
+    });
+
+    reducerManager.addFeatures(feats);
+  },
+});
+
+/**
+ * Environment Initializer used in the feature
+ * providers to register state features
+ */
+const ENVIRONMENT_STATE_PROVIDER: Provider[] = [
+  { provide: FEATURE_STATE_PROVIDER, deps: [], useValue: true },
```

Comment:
The `FEATURE_STATE_PROVIDER` factory will not be executed because of `useValue`. Also, when the factory is used for `InjectionToken`, but `providedIn` is not specified, it fallbacks to `root`.

We can refactor this part as follows:

```ts
export const FEATURE_STATE_PROVIDER = new InjectionToken<void>('@ngrx/store Feature State Provider');

function featureStateProviderFactory(): void {
 // logic is moved from `token.factory` to the standalone function
}

const ENVIRONMENT_STATE_PROVIDER: Provider = [
  { provide: FEATURE_STATE_PROVIDER, useFactory: featureStateProviderFactory },
  {
    provide: ENVIRONMENT_INITIALIZER,
    multi: true,
    useFactory() {
      return () => inject(FEATURE_STATE_PROVIDER);
    },
  },
];
```